### DISCUSSION TITLE
How do I determine the labels for rows in the EquilibriumSensitivity.dudw() matrix output?

### DISCUSSION THREAD

**Discussion by veccp** [reaktoro/reaktoro#380] [Q&A]
### Description

I am working with @bknueven on integrating Reaktoro into WaterTAP. We are trying to get the Jacobian matrix from the dudw() function but are having trouble determining what the rows correspond to (the chemical properties).

### What is your operating system?

None

**Comment by allanleal** [comment]
Hi @veccp , this functionality has not been documented, so I suggest you have a look at the source code of the class [`EquilibriumPredictor`]. There you will learn how a long vector `u` is serialized/deserialized into a [`ChemicalProps`] object. [`EquilibriumPredictor`] can also estimate properties for a given `dw` (a delta of `w`). 

[`EquilibriumPredictor`]: https://github.com/reaktoro/reaktoro/blob/main/Reaktoro/Equilibrium/EquilibriumPredictor.hpp
[`ChemicalProps`]: https://github.com/reaktoro/reaktoro/blob/main/Reaktoro/Core/ChemicalProps.hpp

![image](https://github.com/reaktoro/reaktoro/assets/5825588/28aa195d-04fd-4751-bb6c-dc327087050d)

**Reply by veccp** [reply]
Thanks @allanleal, this has been very helpful. However, [I am confused by](https://github.com/reaktoro/reaktoro/blob/5fa015d178f192e5410b6aa092cc7d533b003938/Reaktoro/Core/ChemicalProps.hpp#L535C1-L536C17) `Vxi`... I am unsure what function this is related to. For example, would I call `speciesPartialMolarVolumes()` for each `phase` to see the values? [This](https://github.com/reaktoro/reaktoro/blob/5fa015d178f192e5410b6aa092cc7d533b003938/Reaktoro/Core/ChemicalProps.cpp#L60) states it should have length `N` (number of species).

**Reply by allanleal** [reply]
Hi @veccp - these are the partial molar volumes of each species in the chemical system organized as an array with size N (number of species). If phase 1 has N1 species and phase 2 has N2 species and N = N1 + N2, then the first N1 entries in the array correspond to the species in phase 1, and the last N2 entries correspond to the species in phase 2.

**Comment by avdudchenko** [comment]
Hello, @allanleal Thank you for all the help. Using provided details we were able to use the acquired jacobian in gray box implementation. (Still working on doing the hessian implementation).  

One question that came up is ability to get the derivatives not just for chemical properties, but also for aqueous properties (specifically saturation indices, pH, etc.). Is there a way to get at these directly? (We can also go about it indirectly, by constructing the constraints that compute those, and getting the jacobian for underlying equilibrium properties, but it would be much simpler if we could get the derivatives directly from reaktoro instead).

Thanks!

**Reply by allanleal** [reply]
Hi @avdudchenko - thanks for the update and I'm glad you're all making good progress. As you note, the Jacobian matrix of aqueous properties is not currently directly available. If you have a `ChemicalProps` object in a state that contains the derivatives of the stored properties with respect to an input variable, say temperature, you can use that `ChemicalProps` object to evaluate the `AqueousProps` object. The temperature derivative calculations will be propagated via automatic differentiation while the aqueous properties are computed. This means you would then get an `AqueousProps` in a state that has temperature derivatives for pH, saturation indices, etc.

**Reply by avdudchenko** [reply]
Thanks. Just to clarify, we would have to write a function to populate the ChemProps object with all the derivatives (e.g there is no way to do that automatically after calling dudw on EquilibriumSensitivity) (at least I did not see an obvious way to pass the calculated derivative matrix into ChemProp object).

**Reply by avdudchenko** [reply]
So, one question here to follow on, I was able to pass in derivatives into a chem-prop object, and evaluate aquous properties useing the updated chem props object. The issue that is sort of obvious is, when we pass in derivative of Temperature with respect to ion concentration ,and its 0, this results in errros like `Unable to interpolate water density at 0 K and 0 Pa because of zero or negative temperature.` but more clearly, its not clear to me which calculations in aqueous props would produce invalid output values simply due to using correlations that expect absolute values (Such as density correlation, and so on). 

It seems to me that some of them would have to be absolute values. - the naive approach seems to effectively do a numeric derivative  (e.g calculate aqueous prop at standard state, and then offset them by the derivative, - evaluate aqueous prop object with differentiated values defined by some small step, and compute the change in props after the fact using the step size of input variable. 

Was this the method you were suggesting @allanleal  or am I missing something in the API?

**Reply by allanleal** [reply]
Hi @avdudchenko - maybe you have already figured out a solution in the meantime. If not, could you please provide a very minimal example demonstrating what you are doing (in terms of ChemicalProps + AqueousProps + derivatives) so that I can carefully check here and evaluate if there is a more appropriate way of achieving it?

**Reply by avdudchenko** [reply]
@allanleal Sorry for taking a while to get back to you. 

(1) Attached is a minimal example (not so minimal I guess =/) - it shows how I would get jacobian for pH from aquous props but it can be used also for scaling tendecies and so on - this actualy works with cyipopt. - this does require applying additional scaling factors to step size to ensure we are not computing steps that are too small, but its not shown in the minimal example. 

(2) On another question, how does one configure reaktoro equilibrium problem for addition of chemicals that alter equilibrium (e.g. adding HCl to water in order to estimate changes in solution pH). There is not really a clear example of how to do that, and obviously we can't simply set H+ species content with amount of HCl added. (the only way I could figure out how to do this (after playing with PhreeqC and figuring out how it does it). is to construct two equilibrium (EQ) problems - the first one solves system equilibrium at initial conditions and key output is total amount of H element (not H+ specie) - this is then used as input into second EQ problem, where to the H amount produced from first EQ problem we add how much HCl is added to the system (increasing total H content. This produces same pH and scaling estimates as PhreeqC (when you run the "reaction" routine), and also can be solved in our platform. Is there a simpler way to do this?

```

import reaktoro as rkt


def _get_rows(state, prop=None):

    species = [species.name() for species in state.system().species()]
    phases = [phase.name() for phase in state.system().phases()]
    print(species, phases)
    # state temperature and pressure
    rows = [("temperature",), ("pressure",)]
    # species amounts * # species
    rows.extend([("speciesAmount", s) for s in species])
    # surface area of reacting surfaces (probably will be here if kinetics are used)
    # temperature of each phase * # phases
    rows.extend([("temperature", p) for p in phases])
    # pressure of each phase * # phases
    rows.extend([("pressure", p) for p in phases])
    # sum speciesAmounts * # phases
    rows.extend([("amount", p) for p in phases])
    # sum speciesMasses * # phases
    rows.extend([("mass", p) for p in phases])
    # mole fractions * # species
    rows.extend([("speciesMoleFraction", s) for s in species])
    # standard molar gibbs free energy * # species
    rows.extend([("speciesStandardGibbsEnergy", s) for s in species])
    # standard molar enthalpy * # species
    rows.extend([("speciesStandardEnthalpy", s) for s in species])
    # standard molar volumes * # species
    rows.extend([("speciesStandardVolume", s) for s in species])
    # temp derivative of standard molar volumes * # species
    rows.extend([("speciesStandardVolumeT", s) for s in species])
    # pres derivative of standard molar volumes * # species
    rows.extend([("speciesStandardVolumeP", s) for s in species])
    # standard isobaric molar heat capacities * # species
    rows.extend([("speciesStandardHeatCapacityConstP", s) for s in species])
    # corrective molar volume * # phases
    rows.extend([("molarVolume", p) for p in phases])
    # temp derivative of corrective molar volume * # phases
    rows.extend([("molarVolumeT", p) for p in phases])
    # pres derivative of corrective molar volume * # phases
    rows.extend([("molarVolumeP", p) for p in phases])
    # mole frac derivative of corrective molar volume * # phases
    rows.extend([("molarVolumeI", s) for s in species])  # not sure what this is
    # corrective gibbs free energy * # phases
    rows.extend([("molarGibbsEnergy", p) for p in phases])
    # corrective molar enthalpy * # phases
    rows.extend([("molarEnthalpy", p) for p in phases])
    # corrective isobaric molar heat capacity * # phases
    rows.extend([("molarHeatCapacityConstP", p) for p in phases])
    # activity coefficients (ln) * # species
    rows.extend([("speciesActivityCoefficientLn", s) for s in species])
    # activities (ln) * # species
    rows.extend([("speciesActivityLn", s) for s in species])
    # chemical potentials * # species
    rows.extend([("speciesChemicalPotential", s) for s in species])
    if prop is not None:
        if isinstance(prop, str):
            jacobian_rows = [
                (idx, row) for idx, row in enumerate(rows) if prop == row[0]
            ]
        else:
            jacobian_rows = [(idx, row) for idx, row in enumerate(rows) if prop == row]
    else:
        jacobian_rows = [(idx, row) for idx, row in enumerate(rows)]

    return jacobian_rows


def generate_jacobian_stream(state, rows, jacobian_matrix, idx, der_step=None):
    data = []
    state_data = []
    _names = []
    for idx_p, prop in rows:

        # print(prop)
        try:
            if len(prop) == 1:
                sval = getattr(state, prop[0])()
            if len(prop) == 2:
                if "Phase" in prop[1]:
                    sval = getattr(state, prop[0])()
                else:
                    sval = getattr(state, prop[0])(prop[1]).val()
        except:
            # print("did not get", prop)
            sval = 0
        # print(prop, sval)
        state_data.append(sval)
        if der_step is not None:
            data.append(sval + jacobian_matrix[idx_p][idx] * der_step)
        else:
            data.append(jacobian_matrix[idx_p][idx])
        _names.append(prop)
    # for idx, nm in enumerate(_names):
    #     print(_names[idx], data[idx], state_data[idx])
    return data, state_data


def transfer_derivatives_to_prop(state, jacobian_matrix, params, input_spec):
    _rows = _get_rows(state)
    jac = []
    aprops = rkt.AqueousProps(state.props())
    base_pH = aprops.pH()

    _state = rkt.ChemicalProps(state.system())
    for param, val in params.items():
        print(param, input_spec[param])
        der_step = 1e-8  # needed to take a small step so we can evalute aquous props
        jac_col, _ = generate_jacobian_stream(
            state.props(), _rows, jacobian_matrix, input_spec[param], der_step=der_step
        )
        # update temp state
        _state.update(jac_col)
        # get aquous props
        aprops = rkt.AqueousProps(_state)
        # print(param, aprops)
        # get jacobian for pH calculated from aquous props
        jac.append(float(aprops.pH() - base_pH) / der_step)
        print(param, aprops.pH(), base_pH, float(aprops.pH() - base_pH))
    # print(aprops.pH(), params, jac)
    return [jac], aprops


def _add_spec_element(specs, param):
    # add a specification to Reaktoro equilibrium problem

    specs.openTo(param)
    idx = specs.addInput(param)
    constraint = rkt.EquationConstraint()
    constraint.id = f"{param}_constraint"
    constraint.fn = lambda props, w: props.speciesAmount(param) - w[idx]
    specs.addConstraint(constraint)


if __name__ == "__main__":

    water_spec = {
        "Na+": 1120 / 22.989769e3,
        "Ca+2": 150 / 40.078e3,
        # "Cl-": 1750 / 35.453e3,
        "HCO3-": 250 / 61.0168e3,
        "H2O": 1000 / 18.01528,
    }

    db = rkt.PhreeqcDatabase("pitzer.dat")
    species = " ".join(list(water_spec.keys()))
    solution = rkt.AqueousPhase(rkt.speciate(species))
    solution.set(rkt.ActivityModelPitzer())

    system = rkt.ChemicalSystem(db, solution)

    state = rkt.ChemicalState(system)

    for ion, values in water_spec.items():
        state.set(ion, values, "mol")
    state.set("H2O", 1000 / 18.01528, "mol")
    state.temperature(25, "celsius")
    state.pressure(1, "atm")
    species = [specie.name() for specie in state.system().elements()]

    print(species)

    specs = rkt.EquilibriumSpecs(system)

    params = {}
    specs.temperature()
    specs.pressure()
    specs.pH()
    specs.charge()
    specs.openTo("Cl-")
    water_spec["H2O"] = 1000 / 18.01528
    for element, con_vals in water_spec.items():
        _add_spec_element(specs, element)

    conditions = rkt.EquilibriumConditions(specs)
    for ion, values in water_spec.items():
        params[ion] = values
        conditions.set(ion, values)
    conditions.pH(7.8)
    params["pH"] = values
    conditions.temperature(298.15)
    params["temperature"] = values
    conditions.pressure(101325)
    params["pressure"] = values
    conditions.charge(0.0)
    input_spec = {}
    for idx, spec in enumerate(specs.namesInputs()):
        if spec == "T":
            input_spec["temperature"] = idx
        elif spec == "P":
            input_spec["pressure"] = idx
        else:
            input_spec[spec] = idx

    solver = rkt.EquilibriumSolver(specs)
    input_names = specs.namesInputs()

    sensitivity = rkt.EquilibriumSensitivity(specs)
    result = solver.solve(state, sensitivity, conditions)
    # get all full jacobian matrix
    _jacobian_matrix = sensitivity.dudw()

    aprops = rkt.AqueousProps(state)
    # get jacobian for pH calculated from aquous props
    aquous_jacobian_matrix, aq_props = transfer_derivatives_to_prop(
        state, _jacobian_matrix, params, input_spec
    )
    print(input_spec)
    print(params)
    print(aprops)
    print(params.keys())
    print(aquous_jacobian_matrix)

```
